### PR TITLE
TFLM: Added VexRISCV optimized DepthwiseConv2d kernel (int8/uint8)

### PR DESCRIPTION
**Pull request by danielyou0230**
This PR adds an optimized DepthwiseConv2D `int8`/`uint8` kernel for VexRISCV as discussed with @advaitjain earlier, see below for details. 

## TL;DR
The optimization
1. Reduces the penalty introduces by excessive memory misses compared to the straightforward implementation
2. No regression on model precision
3. Speedup person_detection (`int8`/`uint8`) by ~5% during inference
4. Requires no special hardware intrinsic instructions or compiler support, just pure C++

## Overview
The kernel is optimized based on the reference kernel in Tensorflow Lite: different from the straightforward implementation, this implementation takes memory layout in TF Lite (`NHWC`) into account, which leverages memory hierarchy to reduce memory miss count, to be more specific, it performs depthwise convolution for every channel in a fixed spatial position (iterate `C`-axis first, then `W`-axis, `H`-axis, and `N`-axis).

This design re-uses memory blocks that have already been cached in faster memory (e.g. L1 cache or L2 cache) and uses a small fixed size buffer of size `k` to store the intermediate values (`int32`) to perform the exact same operation (in this kernel: 2-D depthwise convolution,) thus, **this design has no regression in terms of model precision**.

## Cost of the design
The cost consists of two parts:
1. Uses a small fixed size buffer: memory usage will increase in this layer compared to the original implementation, but this probably will not be an issue (see below for details of how to determine the buffer size `k`)
2. Potentially more memory load when the accumulation is completed: bring buffer for quantization, activation, etc.

## Results
The following results are measured with Digilent Arty A7 with VexRISCV soft-CPU running two versions of person_detection available in the repo, this is not officially supported but basically the same makefile from zephyr_riscv in magic_wand.

#### Original
|       | cycles (M) | Load (M) | Refill (k) |
|-------|:----------:|:--------:|:----------:|
| uint8 |   557.46   |   93.34  |   715.39   |
|  int8 |   503.81   |   87.03  |   561.78   |

#### Optimized
|       | cycles (M) | Load (M) | Refill (k) | 
|:----------:|:--------:|:----------:|:--------:|
|  uint8 |   532.77   |   94.11  |   558.29   |   0.83   |   -21.96   | 157.15 |
|  int8 |   478.38   |   87.62  |   494.90   |   0.67   |   -11.90   | 380.27 |

#### Comparison (delta)
|       | Load | Refill | Miss Penalty (cycles/refill) |
|:----------:|:--------:|:----------:|:--------:|
|  uint8 |  +0.83%   |   -21.96%   | 157.15 |
|  int8 |   +0.67%   |   -11.90%   | 380.27 |

As showed above, with a very marginal amount (< 1%) increment of memory load introduced by the buffer, the optimized kernel reduces the overall memory refill count by at least 10%, which also leads to a speedup (see next section.) Although the measurements are not completely accurate because memory caching is very complicated and it involves a lot of other data in the memory/cache, but the results provide a strong evidence on how improving the memory access pattern impacts the performance on edge devices.

### Inference
|                          |   Original   | Optimized | Speedup |
|--------------------------|:------------:|:---------------:|:----------:|
|                          | # cycles (M) |   # cycles (M)  | (%) |
| person_detection (uint8) |    557.46    |      532.77     |    4.43%   |
| person_detection (int8)  |    503.81    |      478.38     |    5.05%   |

## Details about choice of the buffer size `k`:
The platform I'm mainly testing is Digilent Arty A7 with VexRISC-V soft-CPU. According to the hardware spec specified in [SpinalHDL/VexRiscv](https://github.com/SpinalHDL/VexRiscv), the cache size is 4 KB and the cache block (cache line) size is 32 bytes (search `bytePerLine` in the repo), so there will be 1K blocks available.

In this kernel, the input and output are both `int8` (8 bits, or 1 byte) which means that the system brings 32 `int8` data to the cache each time. To fully utilize the data in this block, we will need 32 accumulator (of type `int32`) for these 32 `int8` data. Thus, a reasonable choice of `k` is 32 and this will take an additional `32 * sizeof(int32) = 128 bytes`, or 4 blocks, this will occupy 4 out of those 1K blocks, which is comparably a small and tolerable overhead.

### CODE REVIEW

**Comment by gbaned** [comment]
Thanks for contributing to TensorFlow Lite Micro.

To keep this process moving along, we'd like to make sure that you have completed the items on this list:
Read the [contributing guidelines for TensorFlow Lite Micro](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md)
Created a [TF Lite Micro Github issue](https://github.com/tensorflow/tensorflow/issues/new?labels=comp%3Amicro&template=70-tflite-micro-issue.md)
Linked to the issue from the PR description

We would like to have a discussion on the Github issue first to determine the best path forward, and then proceed to the PR review.

**Review by advaitjain — APPROVED** [review]
Thanks for the detailed documentation. This PR is good to go.

Having said that, in the coming months we will likely be changing some of the kernel APIs and moving files around as part of improving the infrastructure which might result in this code not building.

We will give you and other optimized kernel maintainers notice and work with you to get the fixes merged.

**Comment by advaitjain** [comment]
Might be worth linking doc/README.md from the top-level readme.
